src/os/LinuxPWM.hxx

```diff
@@ -0,0 +1,174 @@
+                 "/sys/class/pwm/pwmchip%d/pwm%d",
+                 chip_,channel_);
+    }
+    void exportPin()
+    {
+        char exportname[50];
+        strcpy(exportname,pwmdir);
+        strcat(exportname,"/period");
+        FILE *fp = fopen(exportname,"r");
+        if (fp != NULL) {
+            fclose(fp);
+            set_period(1);
+            return;
+        }
+        char *p = strrchr(exportname,'/');
+        *p = '\0';
+        p = strrchr(exportname,'/');
+        p++;
+        strcpy(p,"export");
+        fp = fopen(exportname,"w");
+        fprintf(fp,"%d\n",channel_);
+        fclose(fp);
+        set_period(1);
+    }       
+    void set_period(uint32_t counts) override
+    {
+        char periodname[60];
+        strcpy(periodname,pwmdir);
+        strcat(periodname,"/period");
+        FILE *fp = fopen(periodname,"w");
+        fprintf(fp,"%d\n",counts);
+        fclose(fp);
+        if (counts > 0) {
+            enable();
+        } else {
+            disable();
+        }
+    }
+    uint32_t get_period() override
+    {
+        char periodname[60];
+        uint32_t counts;
+        strcpy(periodname,pwmdir);
+        strcat(periodname,"/period");
+        FILE *fp = fopen(periodname,"r");
+        fscanf(fp,"%d",&counts);
+        fclose(fp);
+        return counts;
+    }
+    void set_duty(uint32_t counts) override
+    {
+        char duty_cyclename[60];
+        strcpy(duty_cyclename,pwmdir);
+        strcat(duty_cyclename,"/duty_cycle");
+        FILE *fp = fopen(duty_cyclename,"w");
+        fprintf(fp,"%d\n",counts);
+        fclose(fp);
+        if (counts > 0) {
+            enable();
+        } else {
+            disable();
+        }
+    }
+    uint32_t get_duty() override
+    {
+        char duty_cyclename[60];
+        uint32_t counts;
+        strcpy(duty_cyclename,pwmdir);
+        strcat(duty_cyclename,"/duty_cycle");
+        FILE *fp = fopen(duty_cyclename,"r");
+        fscanf(fp,"%d",&counts);
+        fclose(fp);
+        return counts;
+    }
+    uint32_t get_period_max() override
+    {
+        return 0xffffffff;
+    }
+    uint32_t get_period_min() override
+    {
+        return 1;
+    }
+    void enable() {
+        char enablename[60];
+        strcpy(enablename,pwmdir);
+        strcat(enablename,"/enable");
+        int fd = open(enablename,O_WRONLY);
+        write(fd,"1\n",2);
+        close(fd);
+    }
+    void disable() {
+        char enablename[60];
+        strcpy(enablename,pwmdir);
+        strcat(enablename,"/enable");
+        int fd = open(enablename,O_WRONLY);
+        write(fd,"0\n",2);
+        close(fd);
+    }
+    bool enabled() {
+        char enablename[60], c;
+        strcpy(enablename,pwmdir);
+        strcat(enablename,"/enable");
+        int fd = open(enablename,O_RDONLY);
+        read(fd,&c,1);
+        close(fd);
+        return (c == '1');
+    }
+private:
+    const uint32_t chip_;
+    const uint32_t channel_;
+    char pwmdir[40];
```

Comment:
this should be pwmdir_, it may also be possible to make this const but I'm not sure.